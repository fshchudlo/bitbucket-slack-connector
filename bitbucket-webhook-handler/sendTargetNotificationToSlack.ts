import * as useCases from "./use-cases";
import { SlackChannel } from "./SlackChannel";
import { BitbucketNotification } from "../bitbucket-payload-types";

export default async function sendTargetNotificationToSlack(payload: BitbucketNotification, pullRequestChannel: SlackChannel, broadcastChannel: SlackChannel = null, defaultChannelParticipants: string[] | null = null) {
    const eventKey = payload.eventKey;

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, pullRequestChannel, defaultChannelParticipants);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, broadcastChannel);
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, pullRequestChannel);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, pullRequestChannel);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, pullRequestChannel);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, pullRequestChannel);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, pullRequestChannel);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, pullRequestChannel);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, pullRequestChannel);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, pullRequestChannel);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, broadcastChannel);
            break;
    }
}

