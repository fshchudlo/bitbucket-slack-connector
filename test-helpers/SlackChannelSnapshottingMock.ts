import {
    SlackChannel,
    BitbucketCommentSnapshot,
    AddBookmarkArguments,
    InviteToChannelArguments,
    KickFromChannelArguments,
    SendMessageArguments,
    SendMessageResponse,
    BitbucketCommentSnapshotInSlackMetadata,
    PullRequestSnapshotInSlackMetadata
} from "../bitbucket-webhook-handler/SlackChannel";
import {
    SNAPSHOT_PULL_REQUEST_STATE_EVENT_TYPE
} from "../bitbucket-webhook-handler/use-cases/helpers/snapshotPullRequestState";
import { SNAPSHOT_COMMENT_STATE_EVENT_TYPE } from "../bitbucket-webhook-handler/use-cases/helpers";
import { SlackChannelInfo } from "../channel-provisioning/SlackChannelFactory";

const messageId = "ABCDE";
export default class SlackChannelSnapshottingMock implements SlackChannel {
    snapshot: {
        addedReactions: any[];
        addedBookmarks: AddBookmarkArguments[];
        closeChannelCalls: string[];
        invitesToChannels: InviteToChannelArguments[];
        kicksFromChannels: KickFromChannelArguments[];
        lookedUpUsers: Array<Array<string>>;
        searchedCommentSnapshots: any[];
        searchedPrOpenedBroadcastMessages: any[];
        sentMessages: SendMessageArguments[];
    };

    constructor() {
        this.snapshot = {
            addedReactions: new Array<any>(),
            addedBookmarks: new Array<AddBookmarkArguments>(),
            closeChannelCalls: new Array<string>(),
            invitesToChannels: new Array<InviteToChannelArguments>(),
            kicksFromChannels: new Array<KickFromChannelArguments>(),
            lookedUpUsers: new Array<Array<string>>(),
            searchedCommentSnapshots: new Array<any>(),
            searchedPrOpenedBroadcastMessages: new Array<any>(),
            sentMessages: new Array<SendMessageArguments>()
        };
    }

    readonly channelInfo: SlackChannelInfo = {
        id: "12345",
        isArchived: false,
        name: "test-channel"
    };

    getSlackUserIds(userEmails: string[]): Promise<string[]> {
        this.snapshot.lookedUpUsers.push(userEmails);
        return Promise.resolve(userEmails);
    }

    addBookmark(options: AddBookmarkArguments): Promise<void> {
        this.snapshot.addedBookmarks.push(options);
        return Promise.resolve();
    }

    inviteToChannel(options: InviteToChannelArguments): Promise<void> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve();
    }

    kickFromChannel(options: KickFromChannelArguments): Promise<void> {
        this.snapshot.kicksFromChannels.push(options);
        return Promise.resolve();
    }

    closeChannel(): Promise<void> {
        this.snapshot.closeChannelCalls.push(this.channelInfo.id);
        return Promise.resolve();
    }

    addReaction(messageId: string, reaction: string): Promise<void> {
        this.snapshot.addedReactions.push({ channelId: this.channelInfo.id, messageId, reaction });
        return Promise.resolve();
    }

    sendMessage(options: SendMessageArguments): Promise<SendMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true, channelId: this.channelInfo.id, messageId: messageId });
    }

    findLatestBitbucketCommentSnapshot(bitbucketCommentId: number | string): Promise<BitbucketCommentSnapshot | null> {
        this.snapshot.searchedCommentSnapshots.push({ channelId: this.channelInfo.id, bitbucketCommentId });

        const snapshot = (<any>this.snapshot.sentMessages).findLast((m: SendMessageArguments) => m.metadata?.eventType === SNAPSHOT_COMMENT_STATE_EVENT_TYPE && m.metadata?.eventPayload?.commentId === bitbucketCommentId.toString());

        if (snapshot) {
            const metadata = <BitbucketCommentSnapshotInSlackMetadata>snapshot.metadata?.eventPayload;
            return Promise.resolve({
                commentId: metadata.commentId,
                commentParentId: metadata.commentParentId,
                threadResolvedDate: metadata.threadResolvedDate,
                taskResolvedDate: metadata.taskResolvedDate,
                severity: metadata.severity,
                slackMessageId: messageId,
                slackThreadId: snapshot.threadId
            });
        }
        return Promise.resolve(null);
    }

    findPROpenedBroadcastMessageId(prCreationDate: Date, pullRequestTraits: PullRequestSnapshotInSlackMetadata): Promise<string | null> {
        this.snapshot.searchedPrOpenedBroadcastMessages.push({
            channelId: this.channelInfo.id,
            prCreationDate,
            pullRequestTraits
        });

        const snapshot = (<any>this.snapshot.sentMessages).findLast((m: SendMessageArguments) => m.metadata?.eventType === SNAPSHOT_PULL_REQUEST_STATE_EVENT_TYPE && m.metadata?.eventPayload?.pullRequestId === pullRequestTraits.pullRequestId && m.metadata?.eventPayload?.projectKey === pullRequestTraits.projectKey && m.metadata?.eventPayload?.repositorySlug === pullRequestTraits.repositorySlug);

        if (snapshot) {
            return Promise.resolve(messageId);
        }
        return Promise.resolve(null);
    }
}
