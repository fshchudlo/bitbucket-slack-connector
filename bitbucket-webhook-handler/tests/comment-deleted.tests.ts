import SlackChannelSnapshottingMock from "../../test-helpers/SlackChannelSnapshottingMock";
import TestPayloadBuilder from "../../test-helpers/TestPayloadBuilder";
import sendTargetNotificationToSlack from "../sendTargetNotificationToSlack";

describe("handleBitbucketWebhook", () => {
    it("Should send message on PR comment deletion", async () => {
        const channelMock = new SlackChannelSnapshottingMock();


        await sendTargetNotificationToSlack(TestPayloadBuilder.pullRequestCommentDeleted(), channelMock);


        expect(channelMock.snapshot).toMatchSnapshot();
    });
});
