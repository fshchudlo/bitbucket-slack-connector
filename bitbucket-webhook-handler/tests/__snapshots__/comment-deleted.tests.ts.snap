// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should send message on PR comment deletion 1`] = `
{
  "addedBookmarks": [],
  "addedReactions": [],
  "closeChannelCalls": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedCommentSnapshots": [
    {
      "bitbucketCommentId": 1,
      "channelId": "12345",
    },
  ],
  "searchedPrOpenedBroadcastMessages": [],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": ":broom: Test Reviewer 1 deleted comment:",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": "> Test comment",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "metadata": {
        "eventPayload": {
          "commentId": "1",
          "commentParentId": undefined,
          "severity": "NORMAL",
          "taskResolvedDate": undefined,
          "threadResolvedDate": undefined,
        },
        "eventType": "bitbucket_comment_snapshot_saved",
      },
      "replyBroadcast": undefined,
      "text": ":broom: Test Reviewer 1 deleted comment:",
      "threadId": undefined,
    },
  ],
}
`;
